### ISSUE TITLE
Color missing, will not compile.

### ISSUE THREAD

**Issue by nanjizal** [ShaharMS/gridge#1]
Color is referenced in ImageTools ( and Image ) try using code without Vision installed.
  
Since Image is fairly common class name it should be in a gridge package.
  
For JS target I have been trying to see if there is a more efficient approach to drawing pixels to a canvas but not yet managed to get it rendering ( added in the Image abstract  'to haxe.io.UInt8Array' and made it @:transient ).
``` Haxe
        var image = new Image( 1024, 768, 0xFFFFFFFF );
        image.fillRect( 100, 100, 300, 200, 0xF0c000 );
        var img: haxe.io.UInt8Array = image;
        var sub = cast( img.subarray( 3, Std.int( image.width*image.height) ), js.lib.Uint8Array );// unsure..
        var data = new js.lib.Uint8ClampedArray( sub.buffer );
        var imageData = new js.html.ImageData( data, image.width, image.height );
        ctx,putImageData(imageData, 0, 0);
```
it would require to store the color differently for js?
```Haxe
	public inline function new(width:Int, height:Int, ?color:Color = 0x00000000) {
		this = new UInt8Array(width * height * 4 + OFFSET);
		this[0] = width & 0xFF;
		this[1] = width >> 8 & 0xFF;
		this[2] = width >> 16 & 0xFF;
		this[3] = width >> 24 & 0xFF;
		var i = 4;
		while (i < this.length) {
			this[i] = color >> 16 & 0xFF;
			this[i + 1] = color >> 8 & 0xFF;
			this[i + 2] = color & 0xFF;
			this[i + 3] = color >> 24 & 0xFF;
			i += 4;
		}
	}

	inline function getColorFromStartingBytePos(position:Int):Color {
		position += OFFSET;
		var c = 0xFF000000;
		c |= this[position]  << 16;
		c |= this[position + 1]<< 8;
		c |= this[position + 2] ;
		c |= this[position + 3]<< 24;
		return c;
	}

	inline function setColorFromStartingBytePos(position:Int, color:Color) {
		position += OFFSET;
		this[position] = color >> 16 & 0xFF;
		this[position + 1] = color >> 8 & 0xFF;
		this[position + 2] = color & 0xFF;
		this[position + 3] = color >> 24 & 0xFF;
		return color;
	}
	```
	thinking maybe have a base abstract that you inherit from that limits access direct and you can typedef for target.
	  
But you may find an alternative for speed is to use an UInt32Array to set color more directly..
var ui32a = new js.lib.Uint32Array( data.buffer ); 
then in for loops you can set the value in one assignment per pixel, rather than accessing each color channel, even if it requires some logic twizzling.
  
Think probably minimal color class should be added to Image when Vision is not used.
  
would like drawTri and fillTri method since you have non AA circle, maybe can provide, but should prob include some canvas optimisations in ImageTools as possibility.  Needs cloneToCanvas method.

**Comment by ShaharMS** [comment]
I agree, I've noticed you're using RGBA format instead of ARGB in your example. The image class uses ARGB

**Comment by nanjizal** [comment]
I think your approach is wrong for working direct in js.
        subarray does not work because it is a view and is wrong size.  To get the correct size you can use slice.
        https://stackoverflow.com/questions/10100798/whats-the-most-straightforward-way-to-copy-an-arraybuffer-object
        ```Haxe
        var img: haxe.io.UInt8Array = image;
        var img2: js.lib.Uint8Array = cast img;
        var sub = img2.slice( 4 );
        var data = new js.lib.Uint8ClampedArray( sub.buffer );
        var imageData = new js.html.ImageData( data, 1024, 768 );
        g.me.putImageData(imageData, 0, 0);
        ```
        My conclusion is you should have different UInt8Array's so that the most effiecient format can be used and have conversion tools. It does not seem ideal to pack the width and height into the abstract as it takes iterations to extract.  Thinking perhaps an Image class with a UInt8Array instance, and width and height.

**Comment by ShaharMS** [comment]
maybe, ill look into that in a couple of days

**Comment by nanjizal** [comment]
Sharhar
I have roughed out an approach here ( committed direct to github ). 
https://github.com/nanjizal/pixelimage/blob/main/src/pixelImage.Pixelimage.hx
But have not debugged tested it yet so maybe very broken but it outlines an approach.

**Comment by nanjizal** [comment]
ok it's ugly as canvas seems to use ABGR  as you said, but now it seems to be finally working.  probably need to consider Image to be in the easiest format for a toolkit to atleast reduce overhead on the image side have added demo code.  The << and >> could be improved, they are verbose for now.
